File: internal/core/backend.go
package core

import "context"

type ScopedKBRef struct {
	Scope string
	ID    string
}

type SearchOptions struct {
	Query      string
	Limit      int
	KBIDs      []ScopedKBRef
	SearchMode string
}

type AddInput struct {
	KBID     string
	Scope    string
	Title    string
	Content  string
	Tags     []string
	Metadata map[string]any
}

type StoreBackend interface {
	Add(context.Context, KnowledgeBase, AddInput) (KnowledgeItem, IngestionResult, IndexStatus, error)
	Search(context.Context, KnowledgeBase, SearchOptions) ([]SearchHit, error)
	GetItem(context.Context, KnowledgeBase, string) (KnowledgeItem, error)
	ListItems(context.Context, KnowledgeBase) ([]KnowledgeItem, error)
	DeleteItem(context.Context, KnowledgeBase, string) error
	SupportsSemantic(KnowledgeBase) bool
}

type IndexOptions struct {
	Rebuild  bool          `json:"rebuild"`
	Progress IndexProgress `json:"-"`
}

// IndexProgressPhase identifies the kind of progress event emitted while
// MaintainIndex runs. Concrete strings are stable; consumers may match on them
// in switch statements.
type IndexProgressPhase string

const (
	IndexProgressPhaseStart        IndexProgressPhase = "start"
	IndexProgressPhaseRebuildReset IndexProgressPhase = "rebuild_reset"
	IndexProgressPhaseIndex        IndexProgressPhase = "index"
	IndexProgressPhaseSkip         IndexProgressPhase = "skip"
	IndexProgressPhaseDeleteOrphan IndexProgressPhase = "delete_orphan"
	IndexProgressPhaseDone         IndexProgressPhase = "done"
)

// IndexProgressEvent is a single observation emitted by MaintainIndex.
// Done is the count of items considered (indexed+skipped) so far; Total is the
// number known up-front (0 if not yet known). Item is the current item id when
// applicable. Message is a human-readable hint, optional.
type IndexProgressEvent struct {
	KBID    string
	Phase   IndexProgressPhase
	Item    string
	Done    int
	Total   int
	Message string
}

type IndexProgress func(IndexProgressEvent)

type IndexResult struct {
	Indexed  int      `json:"indexed"`
	Deleted  int      `json:"deleted"`
	Skipped  int      `json:"skipped"`
	Warnings []string `json:"warnings,omitempty"`
	Errors   []string `json:"errors,omitempty"`
}

type IndexMaintainer interface {
	MaintainIndex(context.Context, KnowledgeBase, IndexOptions) (IndexResult, error)
}

File: knowledger/core/backend.py
from abc import ABC, abstractmethod
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from knowledger.core.types import (
    IngestionResult, IndexStatus, KnowledgeBase, KnowledgeItem, SearchHit
)


@dataclass
class ScopedKBRef:
    scope: str
    id: str


@dataclass
class SearchOptions:
    query: str = ""
    limit: int = 0
    kb_ids: list[ScopedKBRef] = field(default_factory=list)
    search_mode: str = ""


@dataclass
class AddInput:
    kb_id: str
    scope: str
    title: str
    content: str
    tags: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)


class IndexProgressPhase:
    START = "start"
    REBUILD_RESET = "rebuild_reset"
    INDEX = "index"
    SKIP = "skip"
    DELETE_ORPHAN = "delete_orphan"
    DONE = "done"


@dataclass
class IndexProgressEvent:
    kb_id: str
    phase: str
    item: str = ""
    done: int = 0
    total: int = 0
    message: str = ""


IndexProgress = Callable[[IndexProgressEvent], None]


@dataclass
class IndexOptions:
    rebuild: bool = False
    progress: IndexProgress | None = None


@dataclass
class IndexResult:
    indexed: int = 0
    deleted: int = 0
    skipped: int = 0
    warnings: list[str] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)


class StoreBackend(ABC):
    @abstractmethod
    def add(self, kb: KnowledgeBase, inp: AddInput) -> tuple[KnowledgeItem, IngestionResult, IndexStatus]:
        ...

    @abstractmethod
    def search(self, kb: KnowledgeBase, opt: SearchOptions) -> list[SearchHit]:
        ...

    @abstractmethod
    def get_item(self, kb: KnowledgeBase, item_id: str) -> KnowledgeItem:
        ...

    @abstractmethod
    def list_items(self, kb: KnowledgeBase) -> list[KnowledgeItem]:
        ...

    @abstractmethod
    def delete_item(self, kb: KnowledgeBase, item_id: str) -> None:
        ...

    @abstractmethod
    def supports_semantic(self, kb: KnowledgeBase) -> bool:
        ...


class IndexMaintainer(ABC):
    @abstractmethod
    def maintain_index(self, kb: KnowledgeBase, opt: IndexOptions) -> IndexResult:
        ...
